File: public/dsa/searching/search_range/optimal_2/search_range.cpp
#include <bits/stdc++.h>
using namespace std;

int findStart(vector<int> &arr, int key) {
    int low = 0, high = arr.size() - 1, ans = -1;
    while (low <= high) {
        int mid = (low + high) / 2;
        if (arr[mid] == key) {
            ans = mid;
            high = mid - 1;  // keep searching in the left half
        } else if (arr[mid] < key) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return ans;
}

int findEnd(vector<int> &arr, int key) {
    int low = 0, high = arr.size() - 1, ans = -1;
    while (low <= high) {
        int mid = (low + high) / 2;
        if (arr[mid] == key) {
            ans = mid;
            low = mid + 1;  // keep searching in the right half
        } else if (arr[mid] < key) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return ans;
}

vector<int> searchRange(vector<int> &arr, int key) {
    return {findStart(arr, key), findEnd(arr, key)};
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 3, 3, 4, 4, 5};
    int key = 3;
    vector<int> indexes = searchRange(arr, key);
    cout << indexes[0] << " " << indexes[1] << endl;
    return 0;
}
// tests-end

/*output
2 4
*/

File: public/dsa/searching/search_range/optimal_2/search_range.py
from typing import List


class Solution:
    def findStart(self, arr: List[int], key: int) -> int:
        low, high = 0, len(arr) - 1
        ans = -1
        while low <= high:
            mid = (low + high) // 2
            if arr[mid] == key:
                ans = mid
                high = mid - 1  # search left half
            elif arr[mid] < key:
                low = mid + 1
            else:
                high = mid - 1
        return ans

    def findEnd(self, arr: List[int], key: int) -> int:
        low, high = 0, len(arr) - 1
        ans = -1
        while low <= high:
            mid = (low + high) // 2
            if arr[mid] == key:
                ans = mid
                low = mid + 1  # search right half
            elif arr[mid] < key:
                low = mid + 1
            else:
                high = mid - 1
        return ans

    def searchRange(self, arr: List[int], key: int) -> List[int]:
        return [self.findStart(arr, key), self.findEnd(arr, key)]


# tests-start
arr = [1, 2, 3, 3, 3, 4, 4, 5]
key = 3
sol = Solution()
print(sol.searchRange(arr, key))
# tests-end

'''output
[2, 4]
'''
